lib/loglimit/loglimit_test.go

```diff
@@ -0,0 +1,183 @@
+	testCases := []struct {
+		desc             string
+		logSubstrings    []string
+		logsFirstBatch   []string
+		logsSecondsBatch []string
+		logsAssert       func(t *testing.T, loggedFirstBatch, loggedSecondBatch []string)
+	}{
+		{
+			desc: "logs that do not match log substrings are logged",
+			logSubstrings: []string{
+				"A",
+			},
+			logsFirstBatch: []string{
+				"B log 1",
+				"B log 2",
+				"C log 1",
+			},
+			logsSecondsBatch: []string{},
+			logsAssert: func(t *testing.T, loggedFirstBatch, loggedSecondBatch []string) {
+				require.Len(t, loggedFirstBatch, 3)
+				require.Len(t, loggedSecondBatch, 0)
+
+				require.Equal(t, loggedFirstBatch[0], "B log 1")
+				require.Equal(t, loggedFirstBatch[1], "B log 2")
+				require.Equal(t, loggedFirstBatch[2], "C log 1")
+			},
+		},
+		{
+			desc: "logs that match log substrings are deduplicated",
+			logSubstrings: []string{
+				"A",
+				"B",
+			},
+			logsFirstBatch: []string{
+				"A log 1",
+				"B log 1",
+				"B log 2",
+				"B log 3",
+				"C log 1",
+			},
+			logsSecondsBatch: []string{},
+			logsAssert: func(t *testing.T, loggedFirstBatch, loggedSecondBatch []string) {
+				require.Len(t, loggedFirstBatch, 4)
+				require.Len(t, loggedSecondBatch, 0)
+
+				require.Equal(t, loggedFirstBatch[0], "A log 1")
+				require.Equal(t, loggedFirstBatch[1], "B log 1")
+				require.Equal(t, loggedFirstBatch[2], "C log 1")
+				// "A log 1" does not get logged again as the number of occurrences is just 1
+				require.Equal(t, loggedFirstBatch[3], "B log 1 (logs containing \"B\" were seen 3 times in the past minute)")
+			},
+		},
+		{
+			desc: "logs are deduplicated over time windows",
+			logSubstrings: []string{
+				"A",
+				"B",
+			},
+			logsFirstBatch: []string{
+				"A log 1",
+				"B log 1",
+				"B log 2",
+				"B log 3",
+				"C log 1",
+			},
+			logsSecondsBatch: []string{
+				"A log 1",
+				"A log 2",
+				"C log 1",
+				"A log 3",
+				"A log 4",
+			},
+			logsAssert: func(t *testing.T, loggedFirstBatch, loggedSecondBatch []string) {
+				require.Len(t, loggedFirstBatch, 4)
+				require.Len(t, loggedSecondBatch, 3)
+
+				require.Equal(t, loggedFirstBatch[0], "A log 1")
+				require.Equal(t, loggedFirstBatch[1], "B log 1")
+				require.Equal(t, loggedFirstBatch[2], "C log 1")
+				// "A log 1" does not get logged again as the number of occurrences is just 1
+				require.Equal(t, loggedFirstBatch[3], "B log 1 (logs containing \"B\" were seen 3 times in the past minute)")
+
+				require.Equal(t, loggedSecondBatch[0], "A log 1")
+				require.Equal(t, loggedSecondBatch[1], "C log 1")
+				require.Equal(t, loggedSecondBatch[2], "A log 1 (logs containing \"A\" were seen 4 times in the past minute)")
+			},
+		},
+	}
+
+	for _, tc := range testCases {
+		t.Run(tc.desc, func(t *testing.T) {
+			// Create log limiter.
+			// We purposely do not call Run (running the log limiter in a goroutine)
+			// so that we can manually control which actions are performed.
```

Comment:
My 2c is that we should be testing using the public APIs. In this case, I think it's important to call Run() and let it fanout the Log calls.

We can drive the fake clock through the test and either expose a trigger for cleanups, or use a much smaller cleanup window for tests.

Reply:
Done in 0864db932148c68012a899639ae3c617557deefe.